src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs

```diff
@@ -0,0 +1,257 @@
+        /// </summary>
+        private readonly CancellationTokenSource _cancellationTokenSource = new();
+
+        /// <summary>
+        /// Lock controlling mutation of all data (except <see cref="_dismissed"/>) in this indicator, or in any
+        /// sub-scopes. Any read/write of mutable data must be protected by this.
+        /// </summary>
+        public readonly object Gate = new();
+
+        private readonly WpfBackgroundWorkIndicatorFactory _factory;
+        private readonly ITextView _textView;
+        private readonly ITextBuffer _subjectBuffer;
+        private readonly IToolTipPresenter _toolTipPresenter;
+        private readonly ITrackingSpan _trackingSpan;
+        private readonly string _firstDescription;
+        private readonly bool _cancelOnEdit;
+        private readonly bool _cancelOnFocusLost;
+
+        /// <summary>
+        /// Work queue used to batch up UI update and Dispose requests.  A value of <see langword="true"/> means
+        /// just update the tool-tip. A value of <see langword="false"/> means we want to dismiss the tool-tip.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<UIUpdateRequest> _uiUpdateQueue;
+
+        /// <summary>
+        /// Set of scopes we have.  We always start with one (the one created by the initial call to create the work
+        /// indicator). However, the client of the background indicator can add more.
+        /// </summary>
+        private ImmutableArray<BackgroundWorkIndicatorScope> _scopes = ImmutableArray<BackgroundWorkIndicatorScope>.Empty;
+
+        /// <summary>
+        /// If we've been dismissed or not.  Once dismissed, we will close the tool-tip showing information.  This
+        /// field must only be accessed on the UI thread.
+        /// </summary>
+        private bool _dismissed = false;
+
+        private IThreadingContext ThreadingContext => _factory._threadingContext;
+
+        public PropertyCollection Properties { get; } = new();
+        public CancellationToken UserCancellationToken => _cancellationTokenSource.Token;
+        public IEnumerable<IUIThreadOperationScope> Scopes => _scopes;
+
+        public BackgroundWorkIndicatorContext(
+            WpfBackgroundWorkIndicatorFactory factory,
+            ITextView textView,
+            SnapshotSpan applicableToSpan,
+            string firstDescription,
+            bool cancelOnEdit,
+            bool cancelOnFocusLost)
+        {
+            _factory = factory;
+            _textView = textView;
+            _subjectBuffer = applicableToSpan.Snapshot.TextBuffer;
+
+            // Create a tool-tip at the requested position.  Turn off all default behavior for it.  We'll be
+            // controlling everything ourselves.
+            _toolTipPresenter = factory._toolTipPresenterFactory.Create(textView, new ToolTipParameters(
+                trackMouse: false,
+                ignoreBufferChange: true,
```

Comment:
Why turn this off, and handle `cancelOnEdit` ourselves? Is it just not working yet?

Reply:
Primarily because when we handle it ourselves we go through our entire pipeline of steps (including cancel'ing the CTS for our operation).  We don't have an effective way (that i could tell) to know that the tooltip was dismissed due to buffer change.

Reply:
I thought I remember seeing something on the spec, either an event, or a cancellation token that we could use (and created linked token from), but if not, then fair enough. I might have been looking at the V2 doc.